File: Data Structures & Algorithms/subsets/submission-3.java
class Solution {
    
    List<List<Integer>> res;

    public List<List<Integer>> subsets(int[] nums) {
        res = new ArrayList<>();
        List<Integer> curr = new ArrayList<>();
        
        backtrack(0, curr, nums);
        return res;
    }

    public void backtrack(int i, List<Integer> curr, int[] nums) {
        if (i == nums.length) {
            res.add(new ArrayList<>(curr));
            return;
        }

        curr.add(nums[i]);
        backtrack(i + 1, curr, nums);

        curr.remove(curr.size() - 1);
        backtrack(i + 1, curr, nums);
    }
}

File: Data Structures & Algorithms/combination-target-sum-ii/submission-3.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        
        res = []
        curr = []

        def backtrack(i: int, curr: List[int], curr_sum: int) -> None:
            if sum(curr) == target:
                res.append(curr.copy())
                return
            
            if sum(curr) > target or i >= len(candidates):
                return
            
            curr.append(candidates[i])
            backtrack(i + 1, curr, curr_sum + candidates[i])

            curr.pop()
            while i < len(candidates) - 1 and candidates[i] == candidates[i + 1]:
                i += 1
            backtrack(i + 1, curr, curr_sum)
        
        candidates.sort()
        backtrack(0, curr, 0)
        return res

